Tasks/SSH/ssh.ts

```diff
@@ -0,0 +1,201 @@
+/// <reference path="../../definitions/vsts-task-lib.d.ts" />
+
+import path = require('path');
+import tl = require('vsts-task-lib/task');
+import fs = require('fs');
+import Q = require('q');
+var Ssh2Client = require('ssh2').Client;
+var Scp2Client = require('scp2');
+
+function copyScriptToRemoteMachine(scriptFile : string, scpConfig : any) : Q.Promise<string> {
+    var defer = Q.defer<string>();
+
+    Scp2Client.scp(scriptFile, scpConfig, function (err) {
+        if (err) {
+            defer.reject(tl.loc('RemoteCopyFailed', err));
+        } else {
+            tl.debug('Copied script file to remote machine at: ' + scpConfig.path);
+            defer.resolve(scpConfig.path);
+        }
+    });
+
+    return defer.promise;
+}
+
+function setupSshClientConnection(sshConfig: any) : Q.Promise<any> {
+    var defer = Q.defer<any>();
+    var client = new Ssh2Client();
+    client.on('ready', function () {
+        defer.resolve(client);
+    }).on('error', function (err) {
+        defer.reject(err);
+    }).connect(sshConfig);
+    return defer.promise;
+}
+
+function runCommandOnRemoteMachine(sshClient: any, command: string) : Q.Promise<string> {
+    var defer = Q.defer<string>();
+    var stdErrWritten:boolean = false;
+    var stdout: string = '';
+
+    var cmdToRun = command;
+    if(cmdToRun.indexOf(';') > 0) {
+        //multiple commands were passed separated by ;
+        cmdToRun = cmdToRun.replace(/;/g, '\n');
+    }
+    tl.debug('cmdToRun = ' + cmdToRun);
+
+    sshClient.exec(cmdToRun, function(err, stream) {
+        if(err) {
+            defer.reject(tl.loc('RemoteCmdExecutionErr', err))
+        }
+        stream.on('data', function(data) {
+            stdout = stdout.concat(data);
+            if(stdout.endsWith('\n')) {
+                tl._writeLine(stdout);
+                stdout = '';
+            }
+        }).stderr.on('data', function(data) {
+            stdErrWritten = true;
+            tl._writeError(data);
+        })
+        .on('close', function(code, signal) {
+            tl._writeLine(stdout);
+            if(stdErrWritten === true) {
+                defer.reject(tl.loc('RemoteCmdExecutionErr'));
+            } else if(code && code != 0) {
```

Comment:
Is there a chance that there won't be any return code from the execution?  If we always have return code, we probably should relax on the "stdErrWritten === true" check.

Some program writes to stderr but it doesn't mean it failed.  For example, 'curl' writes to stderr for its progress report IIRC.